### ISSUE TITLE
Compute_first_entity_collision for 2D geometries embedded in 3D

### ISSUE THREAD

**Issue by jorgensd** [FEniCS/dolfinx#767]
Following up an issue in the old [ dolfin repo](https://bitbucket.org/fenics-project/dolfin/issues/1098/function-evaluation-returns-wrong-value), I've been able to reproduce the bug for a more dense example in dolfinx. For `Z = 0` the first_entity_collison is able to find the cell given its midpoint. However, it returns cells=[-1] for Z!=0.
```
import os
import sys

from dolfinx.jit import dolfinx_pc, mpi_jit_decorator
import cppimport
from dolfinx.wrappers import get_include_path as pybind_inc

@mpi_jit_decorator
def compile_module():
    cpp_code_header = f"""
    <%
    setup_pybind11(cfg)
    cfg['include_dirs'] += {dolfinx_pc["include_dirs"] + [str(pybind_inc())]}
    cfg['compiler_args'] += {["-D" + dm for dm in dolfinx_pc["define_macros"]]}
    cfg['compiler_args'] += ['-std=c++14']
    cfg['libraries'] += {dolfinx_pc["libraries"]}
    cfg['library_dirs'] += {dolfinx_pc["library_dirs"]}
    %>
    """
    cpp_code ="""
    #include <dolfinx/mesh/Mesh.h>
    #include <dolfinx/mesh/MeshIterator.h>
    #include <pybind11/pybind11.h>
    namespace py = pybind11;
        PYBIND11_MODULE(test_mesh_range, m)
    {
    py::class_<dolfinx::mesh::MeshRange,
    std::shared_ptr<dolfinx::mesh::MeshRange>>(
    m, "MeshRange", "Iterator over entities of a specific dimension.")
    .def(py::init<const dolfinx::mesh::Mesh&, int>())
    .def("__iter__", [](const dolfinx::mesh::MeshRange& r) {
    return py::make_iterator(r.begin(), r.end());
    });
    }
    """
    open(os.path.join(os.getcwd()+"/test123" , "test_mesh_range.cpp"), "w").write(cpp_code_header + cpp_code)

    sys.path.append(os.getcwd()+"/test123")
    return cppimport.imp("test_mesh_range")

module = compile_module()

from dolfinx import Mesh, FunctionSpace, Function, MPI, cpp, geometry,fem, UnitSquareMesh
from dolfinx.cpp.io import permute_cell_ordering, permutation_vtk_to_dolfin
from dolfinx.cpp.mesh import CellType, GhostMode
from dolfinx.io import XDMFFile
import numpy


L, H = 1,1
Z = 0.1
points = numpy.array([[0, 0, 0], [L, 0, 0], [L, H, Z], [0, H, Z],])
cells = numpy.array([[0, 1, 3],
                     [1, 2, 3]])
cells = permute_cell_ordering(cells, permutation_vtk_to_dolfin(CellType.triangle, cells.shape[1]))
mesh = Mesh(MPI.comm_world, CellType.triangle, points, cells, [], GhostMode.none)
cmap = fem.create_coordinate_map(mesh.ufl_domain())
mesh.geometry.coord_mapping = cmap
mesh.create_connectivity_all()

#create DG0 Function Space
DG0 = FunctionSpace(mesh, ("DG", 0))

#create a DG0 function with values
DG0fun = Function(DG0)
vec = numpy.arange(0,DG0fun.vector.local_size)
DG0fun.vector.setArray(vec)

# Get dofmap and cellcenters
dofmap = DG0.dofmap
midpoints =  cpp.mesh.midpoints(mesh, mesh.topology.dim, range(mesh.num_entities(mesh.topology.dim)))
for c in module.MeshRange(mesh, mesh.topology.dim):
    if c.index() == 0:
        continue
    #evaluate function via dof of current cell
    cellvaluefromdof = DG0fun.vector[dofmap.cell_dofs(c.index())[0]]
    #evalute function via generic eval at cell center
    tree = geometry.BoundingBoxTree(mesh, mesh.topology.dim)
    cells = geometry.compute_first_entity_collision(tree, mesh, midpoints[c.index()])
    cellvaluefromeval = DG0fun.eval(midpoints[c.index()], cells)
    print(cellvaluefromdof, cellvaluefromeval)
    assert(cells[0] != -1)
```

**Comment by jorgensd** [comment]
Issue seems to be in ```collides_triangle_point_3d(g.x(v[0]), g.x(v[1]), g.x(v[2]), point);```, which only seems to be working for special cases.

**Comment by jorgensd** [comment]
Further digging reveals that this is due to the fact that the following if-test is entered:
https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfinx/geometry/CollisionPredicates.cpp#L478-L479. 
For the case above, abs(tet_det) = 1e-18, which is not equal to zero. 
Should we replace this test by something like:
```
  if (std::abs(tet_det) > 1e-16)
    return false;
```
which would solve the problem at least for this case.
What do you think @garth-wells ? I think this is the case for several of the collision algorithms.

**Comment by chrisrichardson** [comment]
That line obviously needs changing. In general, I'd like to clean up the collision routines.

**Comment by jorgensd** [comment]
Fixed with #1028, MWE below:
```
from dolfinx import Mesh, FunctionSpace, Function, cpp, geometry,fem
from dolfinx.cpp.io import perm_vtk
from dolfinx.mesh import create_mesh
from dolfinx.cpp.mesh import CellType
from dolfinx.io import XDMFFile
from mpi4py import MPI
import numpy
import ufl

L, H = 1,1
Z = 0.1
points = numpy.array([[0, 0, 0], [L, 0, 0], [L, H, Z], [0, H, Z],])
cells = numpy.array([[0, 1, 3],
                     [1, 2, 3]])
vtk_triangle = perm_vtk(cpp.mesh.CellType.triangle, 3)
cells = cells[:, vtk_triangle]
cell = ufl.Cell("triangle", geometric_dimension=3)
domain =  ufl.Mesh(ufl.VectorElement("Lagrange", cell, 1))
mesh = create_mesh(MPI.COMM_WORLD, cells, points, domain)

#create DG0 Function Space
DG0 = FunctionSpace(mesh, ("DG", 0))

#create a DG0 function with values
DG0fun = Function(DG0)
vec = numpy.arange(3,DG0fun.vector.local_size+3)
DG0fun.vector.setArray(vec)

# Get dofmap and cellcenters
dofmap = DG0.dofmap
l_range = mesh.topology.index_map(mesh.topology.dim).local_range
midpoints =  cpp.mesh.midpoints(mesh, mesh.topology.dim, range(l_range[0], l_range[1]))
for index in range(l_range[0], l_range[1]):
    #evaluate function via dof of current cell
    cellvaluefromdof = DG0fun.vector[dofmap.cell_dofs(index)]
    #evalute function via generic eval at cell center
    tree = geometry.BoundingBoxTree(mesh, mesh.topology.dim)
    cells = geometry.compute_colliding_cells(tree, mesh, midpoints[index])
    cellvaluefromeval = DG0fun.eval(midpoints[index], cells)
    print(cellvaluefromdof, cellvaluefromeval)
```
